app/scripts/views/mixins/flow-events-mixin.js

```diff
@@ -0,0 +1,76 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// Populates the flow model and initialises flow event handlers.
+
+define(function (require, exports, module) {
+  'use strict';
+
+  const Flow = require('models/flow');
+
+  module.exports = {
+    afterRender () {
+      this.flow = new Flow({
+        sentryMetrics: this.sentryMetrics,
+        window: this.window
+      });
+
+      const flowId = this.flow.get('flowId');
+      const flowBegin = this.flow.get('flowBegin');
+
+      this.metrics.setFlowEventMetadata({
+        flowBeginTime: flowBegin,
+        flowId: flowId
+      });
+    },
+
+    // Emit an engage flow event when an input gains focus. However,
+    // don't emit that flow event for the initial autofocus action,
+    // and also don't emit it when the entire window is hidden/shown.
+    afterVisible () {
+      const onInputFocus = () => {
+        this._engageFlowEventsForm();
+      };
+
+      const onWindowFocus = () => {
+        setTimeout(() => {
```

Comment:
> How come the setTimeout here?

Because we're handling the `focus` event, if we immediately attach our listener it will be invoked immediately. It needs to be deferred until after the event has finished. I'll add an explanatory comment to the code.